src/library/scala/concurrent/impl/ExecutionContextImpl.scala

```diff
@@ -53,39 +44,54 @@ private[concurrent] object ExecutionContextImpl {
-          final override def onTermination(exception: Throwable): Unit = deregisterThread()
-          final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T = {
-            var result: T = null.asInstanceOf[T]
-            ForkJoinPool.managedBlock(new ForkJoinPool.ManagedBlocker {
-              @volatile var isdone = false
-              override def block(): Boolean = {
-                result = try {
-                    // When we block, switch out the BlockContext temporarily so that nested blocking does not created N new Threads
-                    BlockContext.withBlockContext(BlockContext.defaultBlockContext) { thunk }
-                  } finally {
-                    isdone = true
-                  }
-
-                true
+    private[this] final def shouldAllocateUnlimited(): Boolean =
+      Thread.currentThread match {
+        case w: WorkerThread if w.isBlocking => false
+        case _ => true
+      }
+
+    override def newThread(runnable: Runnable): Thread =
+      if (shouldAllocateUnlimited()) wire(new Thread(runnable))
+      else if (tryAcquire(1)) wire(new Thread(new Runnable {
+          def run(): Unit = try runnable.run() finally release()
+        }))
+      else null
+
+    override def newThread(fjp: ForkJoinPool): ForkJoinWorkerThread =
+      if (shouldAllocateUnlimited()) wire(new WorkerThread(fjp, null))
+      else if (tryAcquire(1)) wire(new WorkerThread(fjp, this))
+      else null
+
+      final class WorkerThread(final val fjp: ForkJoinPool, final val limiter: Semaphore)
+      extends ForkJoinWorkerThread(fjp) with BlockContext {
+        @volatile var isBlocking: Boolean = false
+        final override def toString: String = s"Worker(isBlocking = $isBlocking, limited = ${limiter ne null})"
+        final override def onTermination(exception: Throwable): Unit = if (limiter ne null) limiter.release()
+        final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T = {
+            isBlocking = true
+            try {
+              if ((limiter ne null) && limiter.availablePermits < 1) {
+                BlockContext.defaultBlockContext.blockOn(thunk)
+              } else {
+                val b = new Block(thunk)
+                ForkJoinPool.managedBlock(b)
+                b.result
               }
-              override def isReleasable = isdone
-            })
-            result
+            } finally {
+              isBlocking = false
+            }
           }
-        })
-      } else null
+      }
+    final class Block[T](thunk: =>T) extends ForkJoinPool.ManagedBlocker {
+      @volatile var done: Boolean = _ // TODO: validate whether @volatile is required or not
+      var result: T = null.asInstanceOf[T]
+      override def isReleasable = done
+      override def block(): Boolean =
+        if (!done) {
+          result = try thunk finally { done = true }
+          true
+        } else true
```

Comment:
always returns true - is it by design?

Reply:
Yes, according to contract set forth by ManagedBlocker.

Reply:
in this case, I would move true out of if block and would delete else branch.

Reply:
done